flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,369 @@
+			hasEmptyBuffer,
+			hasBufferForReleasedChannel,
+			hasBufferForRemovedChannel,
+			normalInputChannelId,
+			releasedInputChannelId);
+
+		ByteBuf[] serializedBuffers = null;
+		ByteBuf mergedBuffer = null;
+
+		try {
+			serializedBuffers = serializeMessages(messages);
+			int[] sizes = getBufferSizes(serializedBuffers);
+			mergedBuffer = mergeBuffers(serializedBuffers);
+
+			List<ByteBuf> partitionedBuffers = partitionBuffer(mergedBuffer, sizes);
+			List<NettyMessage> decodedMessages = decodedMessages(channel, partitionedBuffers);
+
+			verifyDecodedMessages(messages, decodedMessages, normalInputChannelId);
+		} finally {
+			if (serializedBuffers != null) {
+				releaseBuffers(serializedBuffers);
+			}
+
+			if (mergedBuffer != null) {
+				mergedBuffer.release();
+			}
+
+			channel.close();
+		}
+	}
+
+	private List<NettyMessage> createMessageList(
+		boolean hasEmptyBuffer,
+		boolean hasBufferForReleasedChannel,
+		boolean hasBufferForRemovedChannel,
+		InputChannelID normalInputChannelId,
+		InputChannelID releasedInputChannelId) {
+
+		List<NettyMessage> messages = new ArrayList<>();
+
+		Buffer event = createDataBuffer(32);
+		event.tagAsEvent();
+
+		messages.add(new NettyMessage.BufferResponse(createDataBuffer(128), 0, normalInputChannelId, 4));
+		messages.add(new NettyMessage.BufferResponse(createDataBuffer(256), 1, normalInputChannelId, 3));
+
+		if (hasEmptyBuffer) {
+			messages.add(new NettyMessage.BufferResponse(createDataBuffer(0), 1, normalInputChannelId, 2));
+		}
+
+		if (hasBufferForReleasedChannel) {
+			messages.add(new NettyMessage.BufferResponse(createDataBuffer(256), 1, releasedInputChannelId, 3));
+		}
+
+		if (hasBufferForRemovedChannel) {
+			messages.add(new NettyMessage.BufferResponse(createDataBuffer(256), 1, new InputChannelID(), 3));
+		}
+
+		messages.add(new NettyMessage.BufferResponse(event, 2, normalInputChannelId, 4));
+		messages.add(new NettyMessage.ErrorResponse(new RuntimeException("test"), normalInputChannelId));
+		messages.add(new NettyMessage.BufferResponse(createDataBuffer(56), 3, normalInputChannelId, 4));
+
+		return messages;
+	}
+
+	private ByteBuf[] serializeMessages(List<NettyMessage> messages) throws Exception {
+		ByteBuf[] serializedBuffers = new ByteBuf[messages.size()];
+		for (int i = 0; i < messages.size(); ++i) {
+			serializedBuffers[i] = messages.get(i).write(ALLOCATOR);
+		}
+
+		return serializedBuffers;
+	}
+
+	private int[] getBufferSizes(ByteBuf[] buffers) {
+		int[] sizes = new int[buffers.length];
+		for (int i = 0; i < sizes.length; ++i) {
+			sizes[i] = buffers[i].readableBytes();
+		}
+
+		return sizes;
+	}
+
+	private ByteBuf mergeBuffers(ByteBuf[] buffers) {
+		ByteBuf allData = ALLOCATOR.buffer();
+		for (ByteBuf buffer : buffers) {
+			allData.writeBytes(buffer);
+		}
+
+		return allData;
+	}
+
+	private List<ByteBuf> partitionBuffer(ByteBuf buffer, int[] sizes) {
+		// For each group of three messages [A, B, C], we repartition them as three buffers:
```

Comment:
It is better implement a simple way to cut off the buffer slice, with a fixed length, then we do not need to review this complex algorithm which is not very necessary for core function.

Reply:
Simplify the partition, now the buffer is partition with fixed length.